Bazel client error: FMB size limit exceeded

Any call/construction of FMB request on server should be size limited.

```
(18:37:35) ERROR: <redacted>:11:8: Testing <redacted> failed: (Exit 34): Remote Execution Failure:
Unknown: FINDMISSINGBLOBS IS TOO LARGE: 80650 digests are required in one request!
java.io.IOException: com.google.devtools.build.lib.remote.ExecutionStatusException: UNKNOWN: FINDMISSINGBLOBS IS TOO LARGE: 80650 digests are required in one request!
        at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.executeRemotely(GrpcRemoteExecutor.java:235)
        at com.google.devtools.build.lib.remote.RemoteExecutionService.executeRemotely(RemoteExecutionService.java:1496)
        at com.google.devtools.build.lib.remote.RemoteSpawnRunner.lambda$exec$2(RemoteSpawnRunner.java:269)
        at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:244)
        at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:127)
        at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:116)
        at com.google.devtools.build.lib.remote.RemoteSpawnRunner.exec(RemoteSpawnRunner.java:244)
        at com.google.devtools.build.lib.exec.SpawnRunner.execAsync(SpawnRunner.java:301)
        at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:152)
        at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:112)
        at com.google.devtools.build.lib.actions.SpawnStrategy.beginExecution(SpawnStrategy.java:47)
        at com.google.devtools.build.lib.exec.SpawnStrategyResolver.beginExecution(SpawnStrategyResolver.java:64)
        at com.google.devtools.build.lib.exec.StandaloneTestStrategy.beginTestAttempt(StandaloneTestStrategy.java:323)
        at com.google.devtools.build.lib.exec.StandaloneTestStrategy$StandaloneTestRunnerSpawn.beginExecution(StandaloneTestStrategy.java:599)
        at com.google.devtools.build.lib.analysis.test.TestRunnerAction.beginIfNotCancelled(TestRunnerAction.java:991)
        at com.google.devtools.build.lib.analysis.test.TestRunnerAction.beginExecution(TestRunnerAction.java:958)
        at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:1015)
        at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:1007)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$5.execute(SkyframeActionExecutor.java:957)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:1124)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1082)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:160)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:93)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:516)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:827)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:323)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:161)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:571)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:382)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: com.google.devtools.build.lib.remote.ExecutionStatusException: UNKNOWN: FINDMISSINGBLOBS IS TOO LARGE: 80650 digests are required in one request!
        at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.handleStatus(GrpcRemoteExecutor.java:71)
        at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.getOperationResponse(GrpcRemoteExecutor.java:83)
        at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$2(GrpcRemoteExecutor.java:194)
        at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:244)
        at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:127)
        at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:116)
        at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$3(GrpcRemoteExecutor.java:140)
        at com.google.devtools.build.lib.remote.util.Utils.refreshIfUnauthenticated(Utils.java:523)
        at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.executeRemotely(GrpcRemoteExecutor.java:138)
        ... 31 more
```

--- a/src/main/java/build/buildfarm/instance/stub/StubInstance.java
+++ b/src/main/java/build/buildfarm/instance/stub/StubInstance.java
@@ -18,6 +18,7 @@ import static build.buildfarm.common.grpc.Retrier.NO_RETRIES;
 import static build.buildfarm.common.grpc.TracingMetadataUtils.attachMetadataInterceptor;
 import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.allAsList;
 import static com.google.common.util.concurrent.Futures.catching;
 import static com.google.common.util.concurrent.Futures.transform;
 import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
@@ -105,6 +106,7 @@ import com.google.bytestream.ByteStreamGrpc.ByteStreamBlockingStub;
 import com.google.bytestream.ByteStreamGrpc.ByteStreamStub;
 import com.google.bytestream.ByteStreamProto.ReadRequest;
 import com.google.bytestream.ByteStreamProto.ReadResponse;
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Functions;
 import com.google.common.base.Supplier;
 import com.google.common.base.Suppliers;
@@ -163,6 +165,8 @@ public class StubInstance implements Instance {
   private final @Nullable ListeningScheduledExecutorService retryService;
   private boolean isStopped = false;
   private final long maxBatchUpdateBlobsSize = Size.mbToBytes(3);
+
+  @VisibleForTesting long maxRequestSize = Size.mbToBytes(4);
 
   public StubInstance(String name, DigestUtil digestUtil, ManagedChannel channel) {
     this(name, "no-identifier", digestUtil, channel, Durations.fromDays(DEFAULT_DEADLINE_DAYS));
@@ -412,11 +416,16 @@ public class StubInstance implements Instance {
             .setInstanceName(getName())
             .addAllBlobDigests(digests)
             .build();
-    if (request.getSerializedSize() > Size.mbToBytes(4)) {
-      throw new IllegalStateException(
-          String.format(
-              "FINDMISSINGBLOBS IS TOO LARGE: %d digests are required in one request!",
-              request.getBlobDigestsCount()));
+    if (request.getSerializedSize() > maxRequestSize) {
+      // log2n partition for size reduction as needed
+      int partitionSize = (request.getBlobDigestsCount() + 1) / 2;
+      return transform(
+          allAsList(
+              Iterables.transform(
+                  Iterables.partition(digests, partitionSize),
+                  subDigests -> findMissingBlobs(subDigests, requestMetadata))),
+          subMissings -> Iterables.concat(subMissings),
+          directExecutor());
     }
     return transform(
         deadlined(casFutureStub)
@@ -586,6 +595,7 @@ public class StubInstance implements Instance {
       ServerCallStreamObserver<ByteString> blobObserver,
       RequestMetadata requestMetadata) {
     throwIfStopped();
+    checkNotNull(io.grpc.Context.current().getDeadline());
     bsStub
         .get()
         .withInterceptors(attachMetadataInterceptor(requestMetadata))
@@ -886,6 +896,7 @@ public class StubInstance implements Instance {
 
   @Override
   public WorkerListMessage getWorkerList() {
+    checkNotNull(io.grpc.Context.current().getDeadline());
     return workerProfileBlockingStub.get().getWorkerList(WorkerListRequest.newBuilder().build());
   }
 
@@ -897,6 +908,7 @@ public class StubInstance implements Instance {
   @Override
   public CasIndexResults reindexCas() {
     throwIfStopped();
+    checkNotNull(io.grpc.Context.current().getDeadline());
     ReindexCasRequestResults proto =
         adminBlockingStub.get().reindexCas(ReindexCasRequest.newBuilder().build());
     CasIndexResults results = new CasIndexResults();
@@ -910,6 +922,7 @@ public class StubInstance implements Instance {
   @Override
   public void deregisterWorker(String workerName) {
     throwIfStopped();
+    checkNotNull(io.grpc.Context.current().getDeadline());
     adminBlockingStub
         .get()
         .shutDownWorkerGracefully(
@@ -919,6 +932,7 @@ public class StubInstance implements Instance {
   @Override
   public PrepareWorkerForGracefulShutDownRequestResults shutDownWorkerGracefully() {
     throwIfStopped();
+    checkNotNull(io.grpc.Context.current().getDeadline());
     return shutDownWorkerBlockingStub
         .get()
         .prepareWorkerForGracefulShutdown(

--- a/src/test/java/build/buildfarm/instance/stub/StubInstanceTest.java
+++ b/src/test/java/build/buildfarm/instance/stub/StubInstanceTest.java
@@ -64,6 +64,7 @@ import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.util.UUID;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
@@ -98,7 +99,7 @@ public class StubInstanceTest {
     fakeServer.awaitTermination();
   }
 
-  private Instance newStubInstance(String instanceName) {
+  private StubInstance newStubInstance(String instanceName) {
     return new StubInstance(
         instanceName,
         DIGEST_UTIL,
@@ -194,6 +195,38 @@ public class StubInstanceTest {
         ImmutableList.of(Digest.newBuilder().setHash("present").setSizeBytes(1).build());
     assertThat(instance.findMissingBlobs(digests, RequestMetadata.getDefaultInstance()).get())
         .isEmpty();
+    instance.stop();
+  }
+
+  @Test
+  public void findMissingBlobsOverSizeLimitRecombines()
+      throws ExecutionException, InterruptedException {
+    AtomicReference<FindMissingBlobsRequest> reference = new AtomicReference<>();
+    serviceRegistry.addService(
+        new ContentAddressableStorageImplBase() {
+          @Override
+          public void findMissingBlobs(
+              FindMissingBlobsRequest request,
+              StreamObserver<FindMissingBlobsResponse> responseObserver) {
+            reference.set(request);
+            responseObserver.onNext(
+                FindMissingBlobsResponse.newBuilder()
+                    .addAllMissingBlobDigests(request.getBlobDigestsList())
+                    .build());
+            responseObserver.onCompleted();
+          }
+        });
+    StubInstance instance = newStubInstance("findMissingBlobs-test");
+    instance.maxRequestSize = 1024;
+    ImmutableList.Builder<Digest> builder = ImmutableList.builder();
+    // generates digest size * 1024 serialized size at least
+    for (int i = 0; i < 1024; i++) {
+      ByteString content = ByteString.copyFromUtf8("Hello, World! " + UUID.randomUUID());
+      builder.add(DIGEST_UTIL.compute(content));
+    }
+    ImmutableList<Digest> digests = builder.build();
+    assertThat(instance.findMissingBlobs(digests, RequestMetadata.getDefaultInstance()).get())
+        .containsExactlyElementsIn(digests);
     instance.stop();
   }
 

